WikiSentiment/WikiAPIRequests.cs

```diff
@@ -0,0 +1,164 @@
+
+            var articleID = JsonSerializer.Deserialize<Dictionary<string, JsonNode>>(langlinkNode).Keys.First();
+
+            //if article has an english counterpart, it'll have a langlink
+            if (langlinkNode[articleID].AsObject().ContainsKey("langlinks"))
+            {
+                //TODO add conditions
+                return normalizeTitle((string)langlinkNode[articleID]["langlinks"][0]["*"]);
+            }
+            else
+                return "";
+        }
+
+        /// <summary>
+        /// Returns title, if given article is a redirect page. Returns empty if not
+        /// </summary>
+        /// <param name="_client"></param>
+        /// <param name="_title"></param>
+        /// <param name="_country">two letter language code</param>
+        /// <returns>Title of a redirect page, returns empty if not</returns>
+        public static async Task<string> GetRedirect(
+            HttpClient _client, string _title, string _country)
+        {
+            _title = stripSubmenuLink(_title);
+            var request = await _client.GetStringAsync(
+                $"https://{_country}.wikipedia.org/w/api.php?action=parse&" +
+                $"formatversion=2&page={_title}&prop=wikitext&format=json");
+
+            JsonObject jObject = JsonNode.Parse(request).AsObject();
+
+            //validate return json
+            if (!jObject.ContainsKey("parse") ||
+                !jObject["parse"].AsObject().ContainsKey("wikitext"))
+            {
+                return "";
+            }
+
+            //extract title, redirect article text format is "#REDIRECT [[original_title]]"
+            string articleText = ((string)jObject["parse"].AsObject()["wikitext"]).Trim();
+            if (articleText.StartsWith("#") &&
+                articleText.Contains("[[") &&
+                articleText.Contains("]]"))
+            {
+                return normalizeTitle(articleText.Split("[[", 2)[1].Split("]]")[0]);
+            }
+
+            return "";
+        }
+
+        /// <summary>
+        /// Get top pairs of (title, views) for given language
+        /// </summary>
+        /// <param name="_client">Http client with API headers</param>
+        /// <param name="_languageCode">Two letter country code</param>
+        /// <param name="_date">Date for lookup</param>
+        /// <returns>Tuple with articleArray(name, views) of articles</returns>
+        public static async Task<(string title, int views)[]>
+            GetArticleList(HttpClient _client, string _languageCode, DateTime _date)
+        {
+            var articlesString = await _client.GetStringAsync(
+                    "https://wikimedia.org/api/rest_v1/metrics/pageviews/" +
+                    $"top/{_languageCode}.wikipedia.org/all-access/{_date.Year}/{_date.Month:D2}/{_date.Day:D2}");
+
+            var jArticleArray = JsonNode.Parse(articlesString).AsObject()["items"][0]["articles"].AsArray();
```

Comment:
```suggestion
            var jArticleArray = JsonNode.Parse(articlesString).AsObject()["items"][0]["articles"].AsList();
```
This is a specific library, so method might not exist, but in general a `.ToList()` is twice as fast as `.ToArray()`

In general, and in most cases, you should be using [`System.Collections.Generic.List<T>`](https://learn.microsoft.com/en-us/dotnet/api/system.collections.generic.list-1?view=net-8.0) instead of [`T[]`](https://learn.microsoft.com/en-us/dotnet/api/system.array?view=net-8.0)

Reply:
Or [`IEnumerable<T>`](https://learn.microsoft.com/en-us/dotnet/api/system.collections.generic.ienumerable-1?view=net-8.0) if you just need multiple of something and you only need to read it once.

Reply:
There is a suspeciously large number of raw arrays spread around your code